File: javaLang/QuickSortStep.java
package javaLang;

import java.io.*;
import java.util.*;

public class QuickSortStep {
    public static void main(String[] args) throws IOException {
        if (args.length != 3) {
            System.out.println("Usage: java QuickSortStep <filename> <startRow> <endRow>");
            return;
        }

        String filename = args[0];
        int startRow = Integer.parseInt(args[1]);
        int endRow = Integer.parseInt(args[2]);

        List<String> values = new ArrayList<>();

        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            int lineNo = 0;
            while ((line = reader.readLine()) != null) {
                // Use 1-based indexing for row selection
                if (lineNo + 1 >= startRow && lineNo + 1 <= endRow) {
                    values.add(line.trim());
                }
                lineNo++;
            }
        }

        List<String> steps = new ArrayList<>();
        steps.add(convertToSlashFormat(values));

        quickSort(values, 0, values.size() - 1, steps);

        try (BufferedWriter writer = new BufferedWriter(
                new FileWriter("./output/quick_sort_step_" + startRow + "_" + endRow + ".txt"))) {
            for (String step : steps) {
                writer.write(step);
                writer.newLine();
            }
        }
    }

    static void quickSort(List<String> arr, int low, int high, List<String> steps) {
        if (low >= high)
            return;

        int pivotIndex = partition(arr, low, high, steps);
        quickSort(arr, low, pivotIndex - 1, steps);
        quickSort(arr, pivotIndex + 1, high, steps);
    }

    static int partition(List<String> arr, int low, int high, List<String> steps) {
        String pivot = arr.get(high);
        int pivotValue = Integer.parseInt(pivot.split(",")[0]);
        int i = low - 1;

        for (int j = low; j < high; j++) {
            int currentValue = Integer.parseInt(arr.get(j).split(",")[0]);
            if (currentValue < pivotValue) {
                Collections.swap(arr, ++i, j);
            }
        }

        Collections.swap(arr, ++i, high);
        steps.add("pi=" + i + " " + convertToSlashFormat(arr));
        return i;
    }

    static String convertToSlashFormat(List<String> list) {
        List<String> formatted = new ArrayList<>();
        for (String entry : list) {
            String[] parts = entry.split(",");
            formatted.add(parts[0] + "/" + parts[1]);
        }
        return "[" + String.join(", ", formatted) + "]";
    }
}

File: ada_assign/code/pythonLang/quick_sort_step.py
import sys
import os


def convert_to_slash_format(arr):
    """
    Convert array to slash-separated format for logging.
    """
    formatted = [entry.split(",")[0] + "/" + entry.split(",")[1] for entry in arr]
    return "[" + ", ".join(formatted) + "]"


def partition(arr, low, high, steps):
    """
    Partition function for QuickSort with step logging.
    """
    pivot_val = int(arr[high].split(",")[0])
    i = low - 1

    for j in range(low, high):
        curr_val = int(arr[j].split(",")[0])
        if curr_val <= pivot_val:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]

    # Log partition step with full list format
    steps.append(f"pi={i + 1} {convert_to_slash_format(arr)}")

    return i + 1


def quick_sort(arr, low, high, steps):
    """
    QuickSort implementation with step-by-step logging.
    """
    if low < high:
        pi = partition(arr, low, high, steps)
        quick_sort(arr, low, pi - 1, steps)
        quick_sort(arr, pi + 1, high, steps)


def main():
    if len(sys.argv) != 4:
        print("Usage: python quick_sort_step.py <filename> <startRow> <endRow>")
        return

    filename = sys.argv[1]
    start_row = int(sys.argv[2]) - 1  # Adjust to 0-based indexing
    end_row = int(sys.argv[3]) - 1

    data = []

    try:
        # Read only the selected rows
        with open(filename, "r") as file:
            for i, line in enumerate(file):
                if start_row <= i <= end_row:
                    line = line.strip()
                    if line:
                        data.append(line)

        steps = []
        steps.append(convert_to_slash_format(data))  # Initial state

        quick_sort(data, 0, len(data) - 1, steps)

        output_filename = f"./output/quick_sort_step_{start_row + 1}_{end_row + 1}.txt"
        os.makedirs(os.path.dirname(output_filename), exist_ok=True)

        with open(output_filename, "w") as output_file:
            for step in steps:
                output_file.write(step + "\n")

        print(f"Quick Sort steps written to {output_filename}")

    except FileNotFoundError:
        print(f"Error: File '{filename}' not found.")
    except Exception as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    main()
